File: ProiectMpp/ProiectMpp/Repository/TeamDBRepository.cs
using ProiectMpp.Domain;
using log4net.Util;
using Mono.Data.Sqlite;

namespace ProiectMpp.Repository;

public class TeamDBRepository: AbstractDatabaseRepository<int,Team>
{
    public TeamDBRepository(string connectionString) : base(connectionString)
    {
        Load();
        Log.Info("Initializing TeamDBRepository");
    }
     
    public override Team? FindOne(int id)
    {
        Log.Info($"Finding team with id {id}");
        if (Data.ContainsKey(id))
        {
            Log.Info("Team MEM_HIT");
            return Data[id];
        }
        
        SqliteConnection connection = new SqliteConnection(ConnectionString);
        try
        {
            connection.Open();
            using var cmd = connection.CreateCommand();
            cmd.CommandText = "SELECT * FROM Team WHERE id = @id";
            cmd.Parameters.AddWithValue("@id", id);
            using SqliteDataReader reader = cmd.ExecuteReader();
            if (reader.Read())
            {
                Team t = new Team(reader.GetString(0));
                t.Id = reader.GetInt32(1);
                Log.InfoExt("Team found: " + t);
                Data.Add(t.Id, t);
                return t;
            }
            return null;
        }
        catch (Exception e)
        {
                Log.Error(e);
                throw new Exception(e.Message);
        }
        finally
        {
                connection.Close();
        }
        
    }
     
    public new IDictionary<int,Team> FindAll()
    {
        Log.Info("Retrieving all teams");
        return base.FindAll();
        
    }

    protected override void Load()
    {
        Log.Info($"Finding teams");

        SqliteConnection connection = new SqliteConnection(ConnectionString);
        try
        {
            connection.Open();
            using var cmd = connection.CreateCommand();
            cmd.CommandText = "SELECT * FROM Team";
            using SqliteDataReader reader = cmd.ExecuteReader();
            while (reader.Read())
            {
                Team t = new Team(reader.GetString(0));
                t.Id = reader.GetInt32(1);
                Data.Add(t.Id, t);
            }
            Log.InfoExt("Teams found: " + Data.Count.ToString());

        }
        catch (Exception e)
        {
            Log.Error(e);
            throw new Exception(e.Message);
        }
        finally
        {
            connection.Close();
        }
    }
    
    public override Team Save(Team entity)
    {
        Log.Info($"Saving team with id {entity.Id}");
        SqliteConnection connection = new SqliteConnection(ConnectionString);
        try
        {
            connection.Open();
            using var cmd = connection.CreateCommand();
            cmd.CommandText = "INSERT INTO Team (Name) VALUES (@name)";
            cmd.Parameters.AddWithValue("@name", entity.Name);
            cmd.ExecuteNonQuery();
            using (var command = connection.CreateCommand())
            {
                command.CommandText = "SELECT last_insert_rowid();";
                entity.Id = Convert.ToInt32(command.ExecuteScalar());
            }
            Log.InfoExt("Team saved: " + entity);
            Data.Add(entity.Id, entity);
            return entity;
        }
        catch (Exception e)
        {
            Log.Error(e);
            throw new Exception(e.Message);
        }
        finally
        {
            connection.Close();
        }
    }

    public override Team Delete(int id)
    {
        Log.Info($"Deleting team with id {id}");
        SqliteConnection connection = new SqliteConnection(ConnectionString);
        try
        {
            connection.Open();
            using var cmd = connection.CreateCommand();
            cmd.CommandText = "DELETE FROM Team WHERE id = @id";
            cmd.Parameters.AddWithValue("@id", id);
            cmd.ExecuteNonQuery();
            Log.InfoExt("Team deleted: " + id.ToString());
            Team t = Data[id];
            Data.Remove(id);
            return t;
        }
        catch (Exception e)
        {
            Log.Error(e);
            throw new Exception(e.Message);
        }
        finally
        {
            connection.Close();
        }
    }

    public override Team Update(Team entity)
    {
        Log.Info($"Updating team with id {entity.Id}");
        SqliteConnection connection = new SqliteConnection(ConnectionString);
        try
        {
            connection.Open();
            using var cmd = connection.CreateCommand();
            cmd.CommandText = "UPDATE Team SET Name = @name WHERE id = @id";
            cmd.Parameters.AddWithValue("@name", entity.Name);
            cmd.Parameters.AddWithValue("@id", entity.Id);
            cmd.ExecuteNonQuery();
            Log.InfoExt("Team updated: " + entity);
            Team t = Data[entity.Id];
            Data[entity.Id] = entity;
            return t;
        }
        catch (Exception e)
        {
            Log.Error(e);
            throw new Exception(e.Message);
        }
        finally
        {
            connection.Close();
        }
    }
}

File: ProiectSemestru/src/main/java/project/moto/Repository/TeamDBRepository.java
package project.moto.Repository;

import project.moto.Domain.Team;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.util.*;

public class TeamDBRepository extends AbstractDatabaseRepository<Integer, Team> {

    public TeamDBRepository(Properties props) {
        super(props);
        logger.info("Initialized TeamDBRepository with properties: {}", props);
    }

    /**
     * Find the Team with the given id, if it exists in memory in returns instantly, else queries the DB
     * @param aInt -the id of the team to be returned
     * id must not be null
     * @return Team - null if it doesnt exist
     */
    @Override
    public Optional<Team> findOne(Integer aInt) {
        logger.info("Finding Team by ID: {}", aInt);
        if(data.containsKey(aInt))
        {
            logger.traceExit("Team MEMORY_HIT");
            return Optional.of(data.get(aInt));
        }
        Connection con = dbUtils.getConnection();
        try (var preStmt = con.prepareStatement("select * from Team where id = ?")) {
            preStmt.setInt(1, aInt);
            try (var result = preStmt.executeQuery()) {
                if (result.next()) {
                    int id = result.getInt("Id");
                    String name = result.getString("Name");
                    Team team = new Team(name);
                    team.setId(id);
                    logger.traceExit();
                    data.put(id, team);
                    return Optional.of(team);
                }
            }
        } catch (Exception e) {
            logger.error(e.getMessage());
            throw new RuntimeException(e);
        }
        logger.traceExit("Team not found");
        return Optional.empty();
    }

    /**
     * Find all Teams
     * @return Map<Integer,Team> - all Teams
     */
    @Override
    public Map<Integer,Team> findAll() {
        logger.info("Finding all Teams");
        return super.findAll();
    }

    /**
     * Load all Teams from the database
     */
    @Override
    protected void Load(){
        logger.info("Loading Teams from database");
        Connection con = dbUtils.getConnection();
        try (var preStmt = con.prepareStatement("select * from Team")) {
            try (var result = preStmt.executeQuery()) {
                while (result.next()) {
                    int id = result.getInt("id");
                    String name = result.getString("Name");
                    Team team = new Team(name);
                    team.setId(id);
                    data.put(id,team);
                }
            }
        } catch (Exception e) {
            logger.error(e.getMessage());
            throw new RuntimeException(e);
        }
        logger.traceExit("Found {} Teams",data.size());
    }

    /**
     * Save a Team to the database and locally
     * @param entity
     * entity must be not null
     * @return Optional<Team> - null if the Team was not successfully saved
     */
    @Override
    public Optional<Team> save(Team entity) {
        logger.traceEntry("saving team {} ", entity.getId());
        Connection con = dbUtils.getConnection();
        try (var preStmt = con.prepareStatement(
                "insert into Team (Name) values (?)",PreparedStatement.RETURN_GENERATED_KEYS)) {
            preStmt.setString(1, entity.getName());
            var result = preStmt.executeUpdate();
            if (result > 0) {
                try (var generatedKeys = preStmt.getGeneratedKeys()) {
                    if (generatedKeys.next()) {
                        entity.setId(generatedKeys.getInt(1));
                    }
                }
                data.put(entity.getId(), entity);
                logger.traceExit("Team {} saved successfully", entity.getId());
                return Optional.of(entity);
            }
        } catch (Exception e) {
            logger.error(e.getMessage());
            throw new RuntimeException(e);
        }
        logger.traceExit("Team {} not saved", entity.getId());
        return Optional.empty();
    }

    /**
     * Delete a Team from the database and locally
     * @param integer
     * id must be not null
     * @return Optional<Team> - null if the Team was not successfully deleted
     */
    @Override
    public Optional<Team> delete(Integer integer) {
        logger.traceEntry("deleting team {} ", integer);
        Connection con = dbUtils.getConnection();
        try (var preStmt = con.prepareStatement("delete from Team where id = ?")) {
            preStmt.setInt(1, integer);
            preStmt.executeUpdate();
            logger.traceExit("Deleted team {}", integer);
            return Optional.of(data.remove(integer));
        } catch (Exception e) {
            logger.error(e.getMessage());
            throw new RuntimeException(e);
        }
    }

    /**
     * Update a Team in the database and locally
     * @param entity - entity must not be null
     * @return Optional<Team> - null if the Team was not successfully updated
     */
    @Override
    public Optional<Team> update(Team entity) {
        logger.traceEntry("Updating team {} ", entity.getId());
        Connection con = dbUtils.getConnection();
        try (var preStmt = con.prepareStatement("update Team set Name = ? where id = ?")) {
            preStmt.setString(1, entity.getName());
            preStmt.setInt(2, entity.getId());
            preStmt.executeUpdate();
            logger.traceExit("Successfully updated team {} data", entity.getId());
            return Optional.ofNullable(data.put(entity.getId(), entity));

        } catch (Exception e) {
            logger.error(e.getMessage());
            throw new RuntimeException(e);
        }
    }
}
